{python3.pkgs.pybind11, cdrtools, nix}: disable fortify hardening on musl

yu-re-ka opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yuyuyureka committed:
python3.pkgs.pybind11: disable fortify hardening on musl

yuyuyureka committed:
cdrtools: disable fortify hardening on musl

yuyuyureka committed:
nix: disable fortify hardening on musl

risicle commented:
FWIW I think it'd be fine to cause the rebuilds as long as this went to staging, where these packages will almost certainly get rebuilds anyway.

yu-re-ka commented:
Going through staging means I have to wait for the next staging cycle though, and the rebuilds are entirely unnecessary ;)

risicle commented:
Well, mkderivation doesn't seem to like `null` in this context, guessing the only way to avoid a rebuild would be to do a `// optionalAttrs isMusl { hardeningDisable = [...]; }` which is super ugly.

risicle commented:
Incidentally, this made me look into the case where the `nix` build was failing - looks like it fails when code tries to get a function pointer for `vsnprintf`, which is of course impossible with "always_inline". So there are clearly some cases which will never work properly with fortify-headers' approach.

yu-re-ka commented:
> Well, mkderivation doesn't seem to like null in this context

Right, this works for other attributes, but not hardeningDisable.
Changed it back to `lib.optional` and changed the base to staging.
